src/validator.cpp

```diff
@@ -952,63 +962,131 @@ void Validator::ValidatorImpl::validateMathMLElements(const XmlNodePtr &node, co
-                            checkedPairs.push_back(checkPairing);
-
-                            // TODO: validate variable interfaces according to 17.10.8.
-                            // TODO: add check for cyclical connections (17.10.5).
-                            double multiplier = 0.0;
-                            if (!unitsAreEquivalent(model, variable, equivalentVariable, hints, multiplier)) {
-                                auto unitsName = variable->units() == nullptr ? "" : variable->units()->name();
-                                auto equivalentUnitsName = equivalentVariable->units() == nullptr ? "" : equivalentVariable->units()->name();
-                                ErrorPtr err = Error::create();
-                                err->setDescription("Variable '" + variable->name() + "' has units of '" + unitsName + "' and an equivalent variable '" + equivalentVariable->name() + "' with non-matching units of '" + equivalentUnitsName + "'. The mismatch is: " + hints);
-                                err->setModel(model);
-                                err->setKind(Error::Kind::UNITS);
-                                mValidator->addError(err);
-                            }
-
-                            if (equivalentVariable->hasEquivalentVariable(variable)) {
-                                // Check that the equivalent variable has a valid parent component.
-                                auto component2 = std::dynamic_pointer_cast<Component>(equivalentVariable->parent());
-                                if (component2 == nullptr || !component2->hasVariable(equivalentVariable)) {
-                                    ErrorPtr err = Error::create();
-                                    err->setDescription("Variable '" + equivalentVariable->name() + "' is an equivalent variable to '" + variable->name() + "' but '" + equivalentVariable->name() + "' has no parent component.");
-                                    err->setModel(model);
-                                    err->setKind(Error::Kind::CONNECTION);
-                                    mValidator->addError(err);
-                                }
-                            }
-                        }
-                    }
-                }
+void Validator::ValidatorImpl::validateEquivalenceStructure(const VariablePtr &variable)
+{
+    for (size_t index = 0; index < variable->equivalentVariableCount(); ++index) {
+        auto equivalentVariable = variable->equivalentVariable(index);
+        if (equivalentVariable->hasEquivalentVariable(variable)) {
+            auto component = std::dynamic_pointer_cast<Component>(equivalentVariable->parent());
+            if (component == nullptr) {
+                ErrorPtr err = Error::create();
+                err->setDescription("Variable '" + equivalentVariable->name() + "' is an equivalent variable to '" + variable->name() + "' but '" + equivalentVariable->name() + "' has no parent component.");
```

Comment:
Need to add component names to the variable (where poss) so that it's easier for the user to find later

Reply:
Leaving this for the whole error message sweep.